### PR TITLE
[ShaderGraph] Custom Function Node fixes up Function Name to be a valid HLSL identifier

### PR DESCRIPTION
**Pull request by cdxntchou**
### Purpose of this PR
Fix for bugs:
https://fogbugz.unity3d.com/f/cases/1258832/
https://fogbugz.unity3d.com/f/cases/1280188/
https://fogbugz.unity3d.com/f/cases/1184668/

Custom Function Node used to allow you to type any string into the "function name" field, even invalid ones.
Now we fix up any string that isn't an HLSL identifier by converting it to a valid HLSL identifier.

This change also modifies the slot names in the CFN node, so that they are also corrected to be valid HLSL identifiers, and are ensured to be unique when compared against other slots in the CFN node.

This change does not modify existing CFN nodes in any way, but when the function name or slot names are changed in the future, the names will be corrected.

This change also fixes undo/redo for slot renaming.

This PR introduces the new utility function ConvertToValidHLSLIdentifier, with unit tests.
(the old GetHLSLSafeName() has bugs that make it unsafe to use)

It also splits DeduplicateName out from SanitizeName as a more generic useful function, and added unit tests.

### Testing status

Tested entering a variety of different strings for function names, including very long and empty strings.
Tested many different combinations of slot names, including empty strings.
Tested duplicate names for function and slot.
Tested undo/redo.
Tested serialize/save/load/assembly reload.

Added unit tests for ConvertToValidHLSLIdentifier and DeduplicateName.

ShaderGraph tests all pass locally.

**Yamato**: (Select your branch):
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Review by marctem — APPROVED** [review]
Just one question, what you have is fine unless you want to change it.

File: com.unity.shadergraph/Editor/Data/Implementation/NodeUtils.cs [hunk 1]
```diff
@@ -557,6 +558,52 @@ public static string GetHLSLSafeName(string input)
...
+        public static string ConvertToValidHLSLIdentifier(string originalId)
+        {
+            // Converts "  1   var  * q-30 ( 0 ) (1)   " to "_1_var_q_30_0_1"
+            StringBuilder hlslId = new StringBuilder(originalId.Length);
+            bool lastInvalid = false;
+            for (int i = 0; i < originalId.Length; i++)
+            {
+                char c = originalId[i];
+
+                bool isLetter = (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
```
[the remark this answers is not in the document]
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Data/Implementation/NodeUtils.cs:570]: Yes, I had IsLetter and IsDigit originally, but it's not entirely clear what characters those functions accept (and the string is Unicode..), so I opted to do a more explicit implementation to match the HLSL syntax definition.

[the remark this answers is not in the document]
Comment (cdxntchou) [reply com.unity.shadergraph/Editor/Data/Implementation/NodeUtils.cs:570]: IsLetter for example seems to accept letters from other alphabets.  But HLSL spec says A-Z, a-z only.
https://docs.microsoft.com/en-us/dotnet/api/system.char.isletter?view=netcore-3.1

**Comment by cdxntchou** [comment]
Katana all green!

**Review by ghost — APPROVED** [review]
I did test this a bit before scenario test week, I think I can approve it.

**Pull request merged**